pkg/Cpanel/Security/Advisor/Assessors/Symlinks.pm

```diff
@@ -0,0 +1,172 @@
+    my $pid = IPC::Open3::open3( undef, my $out, undef, qw(/sbin/sysctl -a) );
+
+    my %values;
+
+    while ( my $line = readline($out) ) {
+        chomp $line;
+
+        my ( $name, $value ) = ( $line =~ /^(\S+)\s+=\s+(\S+)$/ ) or next;
+
+        $values{$name} = $value;
+    }
+
+    close $out;
+
+    waitpid $pid, 0 or die("Unable to waitpid() on $pid: $!");
+
+    return \%values;
+}
+
+sub generate_advice {
+    my ($self) = @_;
+    $self->_check_for_symlink_kernel_patch;
+
+    return 1;
+}
+
+sub _enforcing_symlink_ownership {
+    my ( $self, $sysctl ) = @_;
+
+    my @vars = qw(
+      kernel.grsecurity.enforce_symlinksifowner
+      fs.enforce_symlinksifowner
+    );
+
+    foreach my $var (@vars) {
+        if ( defined $sysctl->{$var} ) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+sub _symlink_enforcement_gid {
+    my ( $self, $sysctl ) = @_;
+
+    my @vars = qw(
+      kernel.grsecurity.symlinkown_gid
+      fs.symlinkown_gid
+    );
+
+    foreach my $var (@vars) {
+        if ( defined $sysctl->{$var} ) {
+            return int $sysctl->{$var};
+        }
+    }
+
+    return undef;
+}
+
+sub _check_for_symlink_kernel_patch {
+    my ($self) = @_;
+
+    my $security_advisor_obj = $self->{'security_advisor_obj'};
+
+    my $sysctl  = get_sysctl_values();
+    my $sysinfo = Cpanel::GenSysInfo::run();
+
+    #
+    # This test only pertains to RHEL/CentOS 6.
+    #
+    return 1 unless $sysinfo->{'rpm_dist_ver'} == 6;
+
+    #
+    # If a grsecurity kernel is not detected, then we should recommend that
+    # the administrator install one.
+    #
+    unless ( $self->_enforcing_symlink_ownership($sysctl) ) {
+        $self->add_bad_advice(
+            'key'        => q{Symlinks_no_kernel_support_for_ownership_attacks_1},
+            'text'       => ['Kernel does not support the prevention of symlink ownership attacks.'],
+            'suggestion' => ['You do not appear to have any symlink protection enabled through a properly patched kernel on this server, which provides additional protect beyond those solutions employed in userland. Please review the following documentation to learn how to apply this protection.'],
+        );
+
+        return 1;
+    }
+
+    my $gid = $self->_symlink_enforcement_gid($sysctl);
+
+    unless ( defined $gid ) {
+        $self->add_bad_advice(
+            'key'        => q{Symlinks_no_kernel_support_for_ownership_attacks_2},
+            'text'       => ['Kernel does not support the prevention of symlink ownership attacks.'],
+            'suggestion' => ['You do not appear to have any symlink protection enabled through a properly patched kernel on this server, which provides additional protect beyond those solutions employed in userland. Please review the following documentation to learn how to apply this protection.'],
+        );
+
+        return 1;
+    }
+
+    my $shadow = '/etc/shadow';
+    my $dir    = Cpanel::TempFile::get_safe_tmpdir();
```

Comment:
This isn't going to be cleaned up after you're done.  Please consider using the object-oriented interface to make sure that cleanup happens:
```perl
my $tmpobj = Cpanel::TempFile->new;
my $dir = $tmpobj->dir;
```